### DISCUSSION TITLE
How would you implement Date Time based filters

### DISCUSSION THREAD

**Discussion by rugwirobaker** [benbjohnson/wtf#49] [Q&A] [answered]
The way to built queries is very interesting. I wonder how you would approach this if you had a DateTime based filter like:

```
from : "2000-01-01"
to: "2000-01-02"
```

And somehow you had to build a query with `BETWEEN "from" and "to"` as one condition among your others in your where clause.
This is in reference to the magic you did here: https://github.com/benbjohnson/wtf/blob/aab552f94032d424e1588414fd7ef1db9ab6b282/sqlite/dial_membership.go#L144

**Comment by benbjohnson** [comment, marked as answer]
Good question. I usually separate out the two filters. So if it's a filtered on a `Created` timestamp then I would have `CreatedAfter` and `CreatedBefore` fields on my filter:

```go
type WidgetFilter struct {
	CreatedAfter  *time.Time
	CreatedBefore *time.Time
}
```

Then in the SQL construction it would look something like:

```go
if v := filter.CreatedAfter; v != nil {
	where = append(where, "created_at >= ?")
	args = append(args, *v)
}

if v := filter.CreatedBefore; v != nil {
	where = append(where, "created_at < ?")
	args = append(args, *v)
}
```

**Reply by rugwirobaker** [reply]
I suppose this doesn't work for database drivers that use `$1` placeholder instead of `?`, since you would start jumping from `$1` to `$3` if the arg corresponding to `$2` where to be nil.

**Reply by benbjohnson** [reply]
That's a good point. You could use a helper to track the positional arg. e.g.

```go
// posArg returns a SQL positional argument and increments the position.
func posArg(pos *int) string {
	(*pos)++
	return fmt.Sprintf("$%d", *pos)
}
```
